_source/_posts/2021-12-14-antipatterns-secrets-java.adoc

```diff
@@ -0,0 +1,147 @@
+
+Secrets often enter source trees by accident, usually because you temporarily typed one into a config file. When secrets are mixed in with other application properties, they become harder to spot and easier to forget about. Keep secrets out of your source entirely by using an https://12factor.net/config[external configuration] provider like [Hashicorp Vault](https://www.vaultproject.io/[Hashicorp Vault]), using environmental variables, or by keeping secrets in a separate file outside of your source tree (or one that is explicitly ignored in your `.gitignore` file.)
+
+TIP: Check out the https://github.com/paulschwarz/spring-dotenv[spring-dotenv] to use `.env` files with your Spring projects.
+
+
+== 2. Don't conflate encoding with encryption
+
+They are not the same! Many folks are confused about the difference between encoding and encryption; they are distinct operations and cannot be interchanged.
+
+=== Encoding
+
+Data that is _encoded_ is transformed into another format, usually to make transferring between systems possible. An encoded value may or may not be human-readable, but it needs to be treated the same way as the original input source.  The following examples show a string encoded in two common encoding schemes: URL encoding and base64 encoding.
+
+.URL encoded:
+[source,java]
+----
+String text = "Can you still read this text?";
+String encoded = URLEncoder.encode(text, UTF_8);
+// encoded: Can+you+still+read+this+text%3F
+String decoded = URLDecoder.decode(encoded, UTF_8);
+assertThat(decoded, equalTo(decoded));
+----
+
+.Base64 encoded (not human-readable):
+[source,java]
+----
+String text = "Can you still read this text?";
+String encoded = Base64.getEncoder().encodeToString(text.getBytes(UTF_8));
+// encoded: Q2FuIHlvdSBzdGlsbCByZWFkIHRoaXMgdGV4dD8=
+String decoded = new String(Base64.getDecoder().decode(encoded.getBytes(UTF_8)));
+assertThat(decoded, equalTo(decoded));
+----
+
+IMPORTANT: In both cases, the encoded data can easily be reversed, and the encoded string must be treated the same way as plain text.
+
+=== Encryption
+
+In contrast, data that's encrypted can only be read by authorized parties (anyone with the correct cryptographic key). When data is encrypted, the resulting value is indistinguishable from random data. Because of this, encrypted data is often also encoded to make it easier to transport. For example, in the case of JSON Web Encryption (JWE), the JSON payload is encrypted and is then transformed to a base64-encoded string.
+
+NOTE: Encryption is a complex topic and not suitable for a quick example. If you want to see a follow-up post covering common encryption use cases, leave a comment below!
+
+
```

Comment:
You could mention Kubernetes secrets here since by default they're just base64-encoded strings. Sealed secrets is an option here, as discussed in https://developer.okta.com/blog/2021/06/01/kubernetes-spring-boot-jhipster#store-secrets-in-git-with-sealed-secrets-and-kubeseal.

Reply:
```suggestion

If you are using Kubernetes, make sure your secrets are encrypted. By default https://kubernetes.io/docs/concepts/configuration/secret/[Kubernetes Secrets] stores data as base64 encoded strings, as shown above, this isn't any better than plain text. The good news is there are a few alternatives, use HashiCorp's Vault, https://aws.amazon.com/about-aws/whats-new/2021/03/amazon-eks-supports-adding-kms-envelope-encryption-to-existing-clusters/[AWS KMS], or https://dev.to/stack-labs/store-your-kubernetes-secrets-in-git-thanks-to-kubeseal-hello-sealedsecret-2i6h[Sealed Secrets].
```